データベース設計

shunta-nomura opened the pull request:
# What
アプリ開発のためのデータベース設計
(ER図,README)

# WHY
アプリ開発をやりやすく、テーブルの関連を見やすくして、自分やその他の作業者に関係をわかりやすくするため

ER図url
https://gyazo.com/558080ad3f425ca4b6b8c956c8451eb6

shunta-nomura committed:
first_commit

ER図作成ファイルを作成

shunta-nomura committed:
readme記載

shunta-nomura committed:
readme修正

改行がされていなかった

masayaomoto commented on README.md:
```diff
@@ -1,24 +1,59 @@
+|------------------|--------|-------------|
+| nickname         | string | null: false |
+| email            | string | null: false |
+| password         | string | null: false |
```
```
| password         | string | null: false |
```
こちらのカラムはdeviseのgemを使用しての実装になるため、
deviseが予め用意している```encrypted_password```というカラムを用意してあげましょう。

masayaomoto commented on README.md:
```diff
@@ -1,24 +1,59 @@
+| kana_last_name   | string | null: false |
+| year             | string | null: false |
+| month            | string | null: false |
+| day              | string | null: false |
```
```
| year             | string | null: false |
| month            | string | null: false |
| day              | string | null: false |
```
生年月日のカラムはdate型としましょう！
年、月、日、をひとまとめで管理することが可能となります！

masayaomoto commented on README.md:
```diff
@@ -1,24 +1,59 @@
+|Column            |Type         |Options      |
+|------------------|-------------|-------------|
+| name             | string      | null: false |
+| description      | string      | null: false |
```
```
| description      | string      | null: false |
```
こちらのカラムですが、
説明は文章になることも想定しtext型にしましょう！

masayaomoto commented on README.md:
```diff
@@ -1,24 +1,59 @@
+| Status           | string      | null: false |
+| shipping_burden  | string      | null: false |
+| shipping_area    | string      | null: false |
+| shipping_day     | string      | null: false |
```
28~32行目
```
| category         | string      | null: false |
| Status           | string      | null: false |
| shipping_burden  | string      | null: false |
| shipping_area    | string      | null: false |
| shipping_day     | string      | null: false |
```
商品の状態・配送料の負担・発送元の地域・発送までの日数・カテゴリー
に関してのカラムはactive_hashを用いての実装となります。
active_hashを用いたカラムはinteger型にしましょう！
また、カラム名には```_id```を付与しましょう！

masayaomoto commented on README.md:
```diff
@@ -1,24 +1,59 @@
+| shipping_burden  | string      | null: false |
+| shipping_area    | string      | null: false |
+| shipping_day     | string      | null: false |
+| price            | string      | null: false |
```
```
| price            | string      | null: false |
```
こちらのカラムですが、
数値で保存するためinteger型にしましょう！

masayaomoto commented on README.md:
```diff
@@ -1,24 +1,59 @@
+| shipping_day     | string      | null: false |
+| price            | string      | null: false |
+| month            | string      | null: false |
+| user_id          | string      | null: false |
```
```
| user_id          | string      | null: false |
```
こちらは外部キー制約を付与しましょう！

masayaomoto commented on README.md:
```diff
@@ -1,24 +1,59 @@
+| price            | string      | null: false |
+| month            | string      | null: false |
+| user_id          | string      | null: false |
+| image            | activestrage| null: false |
```
```
| image            | activestrage| null: false |
```
こちらのカラムに関しまして、
今回の実装では画像投稿において「active_storage」を使用いたします。
active_storageにて画像データの管理を行いますのでこちらのカラムは不要となります。
削除しておきましょう！

masayaomoto commented on README.md:
```diff
@@ -1,24 +1,59 @@
+belongs_to :user  
+has_one :buy  
+
+## buy
```
現在購入機能に関してのデータはbuyテーブル一つで管理していただいておりますが、
以下の二つのテーブルに分けましょう！
```
配送先住所テーブル（どこに配送するのか住所に関してのデータを管理します）
購入管理テーブル（誰がどの商品を購入したのか購入履歴のデータを管理します）
```

shunta-nomura replied:
typeの設定を行なっていませんでした。すみません。  
・buyテーブルとconsumer_placeの二つのテーブルに元のbuyテーブルを分けました。  
・日付の型はdata型というのを理解し、修正しました。
・active_storageを使用したいカラムには番号をふるintegerを設定することを理解し、設定ました。

以上が変更した点です。
ご確認よろしくお願いします。

yuevo replied:
```
## buy
```
こちらも複数形で記載しましょう！

masayaomoto reviewed (commented):
レビュー依頼ありがとうございます！
コメントしましたので、ご確認をお願いいたします！

shunta-nomura committed:
Revert "readme修正"

This reverts commit 1dd2305e0621d2fb4c80549da14b0328b65261e5.

shunta-nomura committed:
Revert "readme記載"

This reverts commit 9eeef845b6f1b5905c2bd1467dbafb154239c3c1.

shunta-nomura committed:
readmeの記載

er図の修正もあり

shunta-nomura committed:
readme,アクティブストレージ削除

ER図修正

takumi-kodama commented on README.md:
```diff
@@ -1,24 +1,70 @@
+|Column             |Type    |Options      |
+|-------------------|--------|-------------|
+| nickname          | string | null: false |
+| email             | string | null: false |
```
emailカラムには`unique: true`を追加しましょう。

【理由】
悪意あるユーザーが同じemailで複数回登録するのを防ぐためです。

takumi-kodama commented on README.md:
```diff
@@ -1,24 +1,70 @@
+| last_name         | string | null: false |
+| kana_first_name   | string | null: false |
+| kana_last_name    | string | null: false |
+| year              | date   | null: false |
+| month             | date   | null: false |
+| day               | date   | null: false |
```
date型は日付管理を行い、一つにまとめることが可能です。
birthdayなどとし、カラムを統合しましょう。

takumi-kodama commented on README.md:
```diff
@@ -1,24 +1,70 @@
+| Status           | string      | null: false |
+| shipping_burden  | integer     | null: false |
+| shipping_area    | integer     | null: false |
+| shipping_day     | integer     | null: false |
```
Activehashを使用するカラムは命名規則として「モデル名_id」とする必要があります。
例：category_id

そのため、５つのカラム全ての末尾に「_id」を付けましょう。

takumi-kodama commented on README.md:
```diff
@@ -1,24 +1,70 @@
+|------------------|-------------|-------------|
+| name             | string      | null: false |
+| description      | text        | null: false |
+| category         | string      | null: false |
+| Status           | string      | null: false |
```
category,statusもactivehashを使用をします。そのため、integer型へ変更しましょう。

takumi-kodama commented on README.md:
```diff
@@ -1,24 +1,70 @@
-* ...
+|Column            |Type         |Options      |
+|------------------|-------------|-------------|
+| consumer_user_id | integer     | null: false foreign_key: true |
```
外部キーを使用するものはreference型にしましょう。
自動的にカラム名を設定したり、indexを張るなどのメリットがあります。

reference型にする場合はカラム名は`_id`の記述は不要です。
また、カラム名はモデル名と統一させる必要があります。

そのため、46行目は`consumer_user_id`ではなく、`user`としましょう

takumi-kodama commented on README.md:
```diff
@@ -1,24 +1,70 @@
+|Column            |Type         |Options      |
+|------------------|-------------|-------------|
+| consumer_user_id | integer     | null: false foreign_key: true |
+| item_id          | integer     | null: false foreign_key: true |
```
こちらも上記の外部キーに合わせましょう。
`item_id`ではなく、`item`としましょう。

takumi-kodama commented on README.md:
```diff
@@ -1,24 +1,70 @@
+|------------------|-------------|-------------|
+| consumer_user_id | integer     | null: false foreign_key: true |
+| item_id          | integer     | null: false foreign_key: true |
+| consumer_place_id| integer     | null: false foreign_key: true |
```
住所情報の外部キーは不要です。
アソシエーションを用いて情報を取得するためです。

takumi-kodama commented on README.md:
```diff
@@ -1,24 +1,70 @@
+
+### Association
+belongs_to :user  
+has_one :item  
```
`has_one :item`を`belongs_to :item`としましょう。

has_oneは親子の関係によって記述が異なります。
今回はitemモデルが親、buyが子となります。
その場合は親に`has_one`と記述、子には`belongs_to`と記述することで関係性を認識させる必要があるためです。

takumi-kodama commented on README.md:
```diff
@@ -1,24 +1,70 @@
+
+
+### Association 
+has_one :buy
```
こちらも同様です。
購入関連はbuyが親、consumer_placeが子となります。

そのため、70行目は`has_one`ではなく、`belongs_to`を使用しましょう。

takumi-kodama commented on README.md:
```diff
@@ -1,24 +1,70 @@
+| prefecture       | string      | null: false |
+| city_name        | string      | null: false |
+| address          | string      | null: false |
+| building_name    | string      | null: false |
```
建物名は任意のため、`null: false`は不要です。

takumi-kodama reviewed (commented):
実装お疲れ様です。
レビューのご依頼ありがとうございます。

コメントさせていただきました。ご確認をお願いいたします。

shunta-nomura commented:
・reference型を外部キー接続をするときは使用し、カラム名にはモデル名を設定しました。
・date型は一つにまとめられるので、birthdayカラムとしてまとめました。
・emailカラムにはunique: trueを同名emailの登録防止のため追加。
・has_oneの子要素にはbelongs_toにすることを理解し、consumer_placeとbuyのアソシエーションを1つづつbelongs_toに変更
・consumer_placeをbuyモデルから消し、consumer_placeにbuyモデルをreferenceで記載しました。
・activehash型のカラムに_idを割り振り、プルダウン形式の入力カラムをactivehash型に変更しました。

以上の部分を修正いたしました。
コードレビューよろしくお願いします。

shunta-nomura committed:
readme修正

・reference型を外部キー接続をするときは使用し、カラム名にはモデル名を設定しました。
・date型は一つにまとめられるので、birthdayカラムとしてまとめました。
・emailカラムにはunique: trueを同名emailの登録防止のため追加。
・has_oneの子要素にはbelongs_toにすることを理解し、consumer_placeとbuyのアソシエーションを1つづつbelongs_toに変更
・consumer_placeをbuyモデルから消し、consumer_placeにbuyモデルをreferenceで記載しました。
・activehash型のカラムに_idを割り振り、プルダウン形式の入力カラムをactivehash型に変更しました。

masayaomoto commented on README.md:
```diff
@@ -1,24 +1,66 @@
+| name                | string      | null: false |
+| description         | text        | null: false |
+| category_id         | integer     | null: false |
+| Status_id           | integer     | null: false |
```
```| Status_id           | integer     | null: false |```
カラム名は小文字で命名しましょう！
```status_id```

masayaomoto commented on README.md:
```diff
@@ -1,24 +1,66 @@
-* ...
+|Column            |Type         |Options      |
+|------------------|-------------|-------------|
+| consumer_user    | reference   | null: false foreign_key: true |
```
```| consumer_user    | reference   | null: false foreign_key: true |```
こちらusersテーブルの外部キーを設計いたしますので
カラム名は```user```としておきましょう！

masayaomoto commented on README.md:
```diff
@@ -1,24 +1,66 @@
+|Column            |Type         |Options      |
+|------------------|-------------|-------------|
+| post_number      | string      | null: false |
+| prefecture       | string      | null: false |
```
```| prefecture       | string      | null: false |```
都道府県カラムはactive_hashを用いての実装となりますのでinteger型にしましょう！
また、カラム名には```_id```を付与しましょう！

masayaomoto reviewed (commented):
レビュー依頼ありがとうございます！
修正点確認させていただきました。
追加でコメントいたしましたので、ご確認お願いします！

shunta-nomura committed:
readme修正

都道府県カラムを修正しました。
カラム名を2つ修正しました。
itemテーブルのユーザーカラムにnull: falseを追加しました。

shunta-nomura commented:
一部入力忘れなどがありました。すみません。
都道府県カラムを修正しました。
カラム名を2つ修正しました。
itemテーブルのユーザーカラムにnull: falseを追加しました。

以上が修正点です。
コードレビューよろしくお願いします。

yuevo commented on README.md:
```diff
@@ -1,24 +1,66 @@
-# README
+## userテーブル
```
```
## userテーブル
```
テーブル名は`複数形`で記載しましょう！

yuevo commented on README.md:
```diff
@@ -1,24 +1,66 @@
-* System dependencies
 
-* Configuration
+##  itemテーブル
```
```
##  itemテーブル
```
こちら複数形で記載しましょう！

yuevo commented on README.md:
```diff
@@ -1,24 +1,66 @@
+belongs_to :item  
+has_one :consumer_place  　
+
+## consumer_place
```
```
## consumer_place
```
こちらも複数形で記載しましょう！

yuevo reviewed (commented):
レビューのご依頼ありがとうございます！
細かい修正になりますが、４点コメントさせていただいました。
あともう少しなので、頑張っていきましょう！
ご確認のほど、よろしくお願いいたします！

shunta-nomura committed:
readmeテーブル名を複数形に変更

shunta-nomura commented:
READMEとER図のテーブル名を複数形に変更しました。

コードレビューよろしくお願いします。

takumi-kodama reviewed (commented):
実装お疲れ様です。
レビューのご依頼ありがとうございます。

LGTMです！実装お疲れ様でした！
マージ・プルを忘れずに実施していただき、次の機能実装へ移りましょう！
